src/Conversion/ONNXToTOSA/NN/MaxPoolSingleOut.cpp

```diff
@@ -0,0 +1,160 @@
+//
+// Copyright (c) 2022 Advanced Micro Devices, Inc.
+//
+// =============================================================================
+//
+// This file lowers ONNX MaxpoolSingleOut operator to TOSA dialect.
+//
+//===----------------------------------------------------------------------===//
+
+#include "mlir/Dialect/Tosa/IR/TosaOps.h"
+#include "mlir/IR/Attributes.h"
+#include "mlir/IR/BuiltinTypes.h"
+#include "src/Conversion/ONNXToTOSA/DialectBuilder.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSACommon.hpp"
+#include "src/Conversion/ONNXToTOSA/ONNXToTOSALegalizeUtils.hpp"
+#include "src/Dialect/ONNX/ONNXOps/ShapeHelper.hpp"
+#include "llvm/ADT/ArrayRef.h"
+#include <cstdint>
+#include <src/Dialect/Mlir/IndexExpr.hpp>
+
+using namespace mlir;
+
+namespace onnx_mlir {
+
+namespace {
+
+// This calculates the values that need to be added to the padding in order to
+// simulate the ceil mode
+llvm::SmallVector<int64_t> getCeilConstants(llvm::ArrayRef<int64_t> inputShape,
+    const ONNXMaxPoolSingleOutOpShapeHelper &shapeHelper, int64_t ceilMode) {
+  // This avoids having more if statements when creating the padding const.
+  if (ceilMode == 0)
+    return llvm::SmallVector<int64_t>{0, 0};
+
+  SmallVector<int64_t, 4> kernelShapeVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.kernelShape);
+
+  // Get stride and pad vectors.
+  SmallVector<int64_t, 2> stridesVec = shapeHelper.strides;
+  SmallVector<int64_t, 4> padsVec =
+      tosa::createInt64VectorFromIndexExpr(shapeHelper.pads);
+
+  // Check if the idiv_check for the output dimentions in
+  // https://www.mlplatform.org/tosa/tosa_spec.html#_max_pool2d has no
+  // remainder. If it has a remainder, we add size(stride) to the end of the
+  // padding dimension to get one dimension up. Height and width need to have
+  // seperate values.
+  int64_t xAxis = 0;
+  if ((inputShape[2] + padsVec[0] + padsVec[2] - kernelShapeVec[0]) %
+      stridesVec[0])
+    xAxis = stridesVec[0];
+
+  int64_t yAxis = 0;
+  if ((inputShape[3] + padsVec[1] + padsVec[3] - kernelShapeVec[1]) %
+      stridesVec[1])
+    yAxis = stridesVec[1];
+
+  return llvm::SmallVector<int64_t>{xAxis, yAxis};
+}
+
+class ONNXMaxPoolSingleOutOpLoweringToTOSA : public ConversionPattern {
+public:
+  ONNXMaxPoolSingleOutOpLoweringToTOSA(MLIRContext *ctx)
+      : ConversionPattern(ONNXMaxPoolSingleOutOp::getOperationName(), 1, ctx) {}
+
+  using OpAdaptor = typename ONNXMaxPoolSingleOutOp::Adaptor;
+  LogicalResult matchAndRewrite(Operation *op, ArrayRef<Value> operands,
+      ConversionPatternRewriter &rewriter) const final {
+    auto maxpoolOp = llvm::cast<ONNXMaxPoolSingleOutOp>(op);
+    Location loc = op->getLoc();
+    OpAdaptor adaptor(operands, op->getAttrDictionary());
+    // Get shape.
+    IndexExprBuilderForTosa createTosaIE(rewriter, loc);
+    ONNXMaxPoolSingleOutOpShapeHelper shapeHelper(op, operands, &createTosaIE);
```

Comment:
FYI, if you need it in the future.

If you were to replace `operands` by `{}`, the shape helper will see the original ONNX ops.

Reply:
Thank you, this can be really helpful in some scenarios!